### PR TITLE
Added extra_specs option (Fixes #3272)

### LINKED ISSUE #3272: Ansible Openstack module in OS_NOVA_FLAVOR
##### ISSUE TYPE

Feature Idea
##### COMPONENT NAME

os_nova_flavor module
##### ANSIBLE VERSION

N/A
##### SUMMARY

There is no argument right now to pass extra spec to flavor. These are very important to for creation of NUMA, CPU PINNING , SRIOV etc . 

Openstack support it but ansible does not support natively . 

Regards
Arif Khan

### PR DESCRIPTION
**Pull request by bingoarun**
##### ISSUE TYPE
- Bugfix Pull Request
##### COMPONENT NAME

os_nova_flavor
##### ANSIBLE VERSION

```
# ansible --version
ansible 2.0.0.0
```
##### SUMMARY

Added extra_specs option in flavor creation

Added support for including an optional extra_specs option while creating flavors. Fixes bug #3272 

```
# cat create_flavor.yml 

---
- hosts: localhost
  tasks:
   - name: 
     os_nova_flavor_new:
      cloud=packstack
      state=present
      name=mynewflavor
      ram=1234
      vcpus=1
      disk=10
      ephemeral=10
      extra_specs="quota:cpu_quota=10000,quota:disk_write_bytes_sec=10485760"
```

```
# ansible-playbook -v create_flavor.yml --connection=local
No config file found; using defaults
 [WARNING]: provided hosts list is empty, only localhost is available


PLAY ***************************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [None] ********************************************************************
changed: [localhost] => {"changed": true, "flavor": {"OS-FLV-DISABLED:disabled": false, "OS-FLV-EXT-DATA:ephemeral": 10, "disk": 10, "ephemeral": 10, "extra_specs": {}, "id": "d192b4e8-bb46-450c-892a-9fc95989c383", "is_public": true, "name": "mynewflavor", "os-flavor-access:is_public": true, "ram": 1234, "request_ids": ["req-1c98161a-9dc5-45ab-acd9-f416ab9298f4", "req-535c3730-def9-4832-9fd6-92764cc29c14"], "rxtx_factor": 1.0, "swap": "", "vcpus": 1, "x_openstack_request_ids": ["req-1c98161a-9dc5-45ab-acd9-f416ab9298f4", "req-535c3730-def9-4832-9fd6-92764cc29c14"]}, "id": "d192b4e8-bb46-450c-892a-9fc95989c383"}

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0 
```

```
# ansible-playbook -v create_flavor.yml --connection=local
No config file found; using defaults
 [WARNING]: provided hosts list is empty, only localhost is available


PLAY ***************************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [None] ********************************************************************
changed: [localhost] => {"changed": true, "flavor": {"OS-FLV-DISABLED:disabled": false, "OS-FLV-EXT-DATA:ephemeral": 10, "disk": 10, "ephemeral": 10, "extra_specs": {}, "id": "d192b4e8-bb46-450c-892a-9fc95989c383", "is_public": true, "name": "mynewflavor", "os-flavor-access:is_public": true, "ram": 1234, "request_ids": ["req-1c98161a-9dc5-45ab-acd9-f416ab9298f4", "req-535c3730-def9-4832-9fd6-92764cc29c14"], "rxtx_factor": 1.0, "swap": "", "vcpus": 1, "x_openstack_request_ids": ["req-1c98161a-9dc5-45ab-acd9-f416ab9298f4", "req-535c3730-def9-4832-9fd6-92764cc29c14"]}, "id": "d192b4e8-bb46-450c-892a-9fc95989c383"}

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0 
```

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @bingoarunprasath for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by bingoarun** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @bingoarunprasath. To the current maintainers, @emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/openstack/os_nova_flavor.py [hunk 1]
```diff
@@ -206,6 +216,12 @@ def main():
 
         if state == 'present':
             if not flavor:
+                if type(module.params['extra_specs']) is str:
+                    extra_specs = {}
+                    for kv_str in module.params['extra_specs'].split(","):
+                        k, v = kv_str.split("=")
+                        extra_specs[k] = v
+                    module.params['extra_specs'] = extra_specs
```
Comment (omgjlk) [inline cloud/openstack/os_nova_flavor.py]: From reading what the Ansible module helpers do with a type of dict, this bit should be done for you. Have you tested this out without putting in the string conversion code?

File: cloud/openstack/os_nova_flavor.py [hunk 2]
```diff
@@ -217,6 +233,7 @@ def main():
                     rxtx_factor=module.params['rxtx_factor'],
                     is_public=module.params['is_public']
                 )
+                cloud.set_flavor_specs(flavor['id'],module.params['extra_specs'])
```
Comment (omgjlk) [inline cloud/openstack/os_nova_flavor.py]: Does this work if the value of `module.params['extra_specs']` is `None`?

**Comment by omgjlk** [comment]
Since this module is missing documentation on the return value, can you share whether the extra specs are now returned when a flavor is created?

**Comment by bingoarun** [comment]
Hi @j2sol , Thanks for reviewing this. 

Yes, you are right. String values are automatically converted to dict. So I have removed the string conversion code.

Regarding return values, when a flavor is created, it returns the object with empty 'extra_specs'. Then 'set_flavor_specs' function is called to set the extra_specs value. This 'set_flavor_specs' doesn't return any value.

**Comment by rcarrillocruz** [comment]
Hi @bingoarunprasath 

Change looks ok, however can you please squash this PR into one commit?

Thanks

**Comment by rcarrillocruz** [comment]
needs_revision

**Comment by bingoarun** [comment]
Hi @rcarrillocruz : I have squashed it to https://github.com/ansible/ansible-modules-core/pull/4068/commits/5a326eda723f8e03717fd31182f238b2f5f2908d

**Comment by bingoarun** [comment]
ready_for_review

**Comment by rcarrillocruz** [comment]
I'm afraid you haven't squashed this PR into one commit, as you can see in the Commits tab it shows 6.

Check it out this link on how to use git rebase to squash multiple changes into one:

http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

needs_revision

**Comment by bingoarun** [comment]
@rcarrillocruz  :  Oops, thanks for notifying. Fixed it now.

**Comment by bingoarun** [comment]
ready_for_review

**Comment by emonty** [comment]
shipit

**Comment by gundalow** [comment]
Hi,

Looks like this branch needs updating to fix the Shippable issue, please see this guide on how to do that cleanly.

http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

**Comment by gregdek** [comment]
Thanks again to @bingoarunprasath for this PR, and thanks @emonty, @shrews, @juliakreger, @j2sol, @rcarrillocruz for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by omgjlk** [comment]
@bingoarunprasath thanks for explaining. It'd be good in a future change to include extra spec data in the return data of the module.

**Comment by gregdek** [comment]
Thanks @bingoarun for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@bingoarun A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Pull request closed without merging**